File: backend/app/offline_payments.py
"""Offline POS payment safety — cash default; no fabricated provider success (§18)."""

from __future__ import annotations

from fastapi import HTTPException

from app.rbac import has_permission

# Safe offline without supervisor acknowledgment.
OFFLINE_CASH_METHODS = frozenset({"cash"})

# External / provider rails — never claim live approval when queued offline.
OFFLINE_PROVIDER_METHODS = frozenset(
    {
        "card",
        "wallet",
        "bank_transfer",
        "mobile_money",
        "momo",
        "online",
    }
)

OFFLINE_CREDIT_METHOD = "credit"

SUPERVISOR_ROLES = frozenset(
    {
        "store_manager",
        "company_admin",
        "super_admin",
        "sales_officer",
        "accountant",
        "tenant_owner",
        "tenant_admin",
    }
)


def normalize_payment_method(method: str | None) -> str:
    return (method or "cash").strip().lower()


def sale_payment_methods(payload: dict) -> list[str]:
    payments = payload.get("payments")
    if isinstance(payments, list) and payments:
        methods: list[str] = []
        for row in payments:
            if not isinstance(row, dict):
                continue
            methods.append(normalize_payment_method(str(row.get("payment_method") or "cash")))
        return methods or [normalize_payment_method(str(payload.get("payment_method") or "cash"))]
    return [normalize_payment_method(str(payload.get("payment_method") or "cash"))]


def _offline_meta(payload: dict) -> dict:
    nested = payload.get("payload")
    if isinstance(nested, dict):
        return nested
    return {}


def _supervisor_ack(payload: dict) -> tuple[bool, str]:
    meta = _offline_meta(payload)
    ack = bool(payload.get("offline_supervisor_ack") or meta.get("offline_supervisor_ack"))
    reason = str(
        payload.get("offline_supervisor_reason") or meta.get("offline_supervisor_reason") or ""
    ).strip()
    return ack, reason


def _credit_cached_ack(payload: dict) -> bool:
    meta = _offline_meta(payload)
    return bool(payload.get("offline_credit_cached_ack") or meta.get("offline_credit_cached_ack"))


def is_supervisor_offline_actor(claims: dict) -> bool:
    role = (claims.get("role") or "cashier").strip().lower()
    if role in SUPERVISOR_ROLES:
        return True
    overrides = claims.get("permissions") if isinstance(claims.get("permissions"), dict) else None
    for module in ("credit", "expenses", "purchasing"):
        if has_permission(role, module, "approve", overrides=overrides):
            return True
    return False


def validate_offline_pos_sale_payments(payload: dict, *, claims: dict) -> None:
    """Reject unsafe offline payment combinations before record_pos_sale runs."""
    if not isinstance(payload, dict):
        raise HTTPException(status_code=400, detail="invalid pos_sale payload")

    methods = sale_payment_methods(payload)
    ack, reason = _supervisor_ack(payload)

    for method in methods:
        if method in OFFLINE_CASH_METHODS:
            continue

        if method == OFFLINE_CREDIT_METHOD:
            if not (payload.get("party_id") or "").strip():
                raise HTTPException(
                    status_code=400,
                    detail=(
                        "OFFLINE_CREDIT_BLOCKED: Offline credit requires a registered customer "
                        "(party_id)"
                    ),
                )
            if not _credit_cached_ack(payload):
                raise HTTPException(
                    status_code=400,
                    detail=(
                        "OFFLINE_CREDIT_BLOCKED: Offline credit requires cached customer "
                        "acknowledgment (revalidated on sync)"
                    ),
                )
            role = (claims.get("role") or "cashier").strip().lower()
            overrides = (
                claims.get("permissions") if isinstance(claims.get("permissions"), dict) else None
            )
            if not (
                has_permission(role, "credit", "read", overrides=overrides)
                or has_permission(role, "credit", "write", overrides=overrides)
            ):
                raise HTTPException(
                    status_code=400,
                    detail="OFFLINE_CREDIT_BLOCKED: Missing credit permission for offline credit sale",
                )
            continue

        if method in OFFLINE_PROVIDER_METHODS or method not in OFFLINE_CASH_METHODS:
            if not ack:
                raise HTTPException(
                    status_code=400,
                    detail=(
                        "OFFLINE_PAYMENT_BLOCKED: "
                        f"{method} payments cannot be queued offline without supervisor "
                        "acknowledgment — provider success is not verified offline"
                    ),
                )
            if len(reason) < 3:
                raise HTTPException(
                    status_code=400,
                    detail=(
                        "OFFLINE_PAYMENT_BLOCKED: Supervisor acknowledgment reason required "
                        "(min 3 characters) for offline provider payments"
                    ),
                )

File: frontend/lib/offlinePayments.ts
/**
 * Offline POS payment safety (§18) — cash default; no fabricated provider success.
 */

export const OFFLINE_CASH_METHODS = new Set(['cash']);

export const OFFLINE_PROVIDER_METHODS = new Set([
  'card',
  'wallet',
  'bank_transfer',
  'mobile_money',
  'momo',
  'online',
]);

export const OFFLINE_CREDIT_METHOD = 'credit';

export const OFFLINE_CREDIT_CACHED_NOTE = 'Cached customer credit data — revalidated on sync';

const SUPERVISOR_ROLES = new Set([
  'store_manager',
  'company_admin',
  'super_admin',
  'sales_officer',
  'accountant',
  'tenant_owner',
  'tenant_admin',
]);

export type OfflinePaymentContext = {
  role: string;
  permissions: Record<string, string[]> | null;
  customers: Array<{
    id: string;
    credit_limit?: number;
    balance?: number;
  }>;
  customerId: string;
};

export function normalizePaymentMethod(method: string | null | undefined): string {
  return (method || 'cash').trim().toLowerCase();
}

export function salePaymentMethods(body: Record<string, unknown>): string[] {
  const payments = body.payments;
  if (Array.isArray(payments) && payments.length) {
    return payments.map((p) =>
      normalizePaymentMethod(
        typeof p === 'object' && p && 'payment_method' in p
          ? String((p as { payment_method?: string }).payment_method || 'cash')
          : 'cash',
      ),
    );
  }
  return [normalizePaymentMethod(String(body.payment_method || 'cash'))];
}

export function canSupervisorOfflinePayment(
  role: string,
  permissions: Record<string, string[]> | null,
): boolean {
  const roleNorm = (role || 'cashier').trim().toLowerCase();
  if (SUPERVISOR_ROLES.has(roleNorm)) return true;
  if (!permissions) return false;
  return ['credit', 'expenses', 'purchasing'].some((module) =>
    (permissions[module] || []).includes('approve'),
  );
}

export function canOfflineCredit(
  permissions: Record<string, string[]> | null,
): boolean {
  if (!permissions) return false;
  if (permissions['*']?.includes('*')) return true;
  const credit = permissions.credit || [];
  return credit.includes('*') || credit.includes('read') || credit.includes('write');
}

export function hasCachedCustomerCredit(
  ctx: OfflinePaymentContext,
): boolean {
  if (!ctx.customerId) return false;
  const customer = ctx.customers.find((c) => c.id === ctx.customerId);
  if (!customer) return false;
  return customer.credit_limit != null || customer.balance != null;
}

export type OfflinePaymentPrep = {
  payloadMeta: Record<string, unknown>;
  userMessage?: string;
  blocked?: boolean;
  error?: string;
  requiresSupervisorPrompt?: boolean;
};

export function prepareOfflineSalePayments(
  body: Record<string, unknown>,
  ctx: OfflinePaymentContext,
  opts?: { supervisorReason?: string },
): OfflinePaymentPrep {
  const methods = salePaymentMethods(body);
  const payloadMeta: Record<string, unknown> = {
    offline_queued_at: new Date().toISOString(),
  };
  const isSupervisor = canSupervisorOfflinePayment(ctx.role, ctx.permissions);

  for (const method of methods) {
    if (OFFLINE_CASH_METHODS.has(method)) continue;

    if (method === OFFLINE_CREDIT_METHOD) {
      if (!ctx.customerId) {
        return {
          payloadMeta,
          blocked: true,
          error: 'Offline credit requires a registered customer',
        };
      }
      if (!hasCachedCustomerCredit(ctx)) {
        return {
          payloadMeta,
          blocked: true,
          error: 'Offline credit requires cached customer credit data from the last online load',
        };
      }
      if (!canOfflineCredit(ctx.permissions)) {
        return {
          payloadMeta,
          blocked: true,
          error: 'Missing credit permission for offline credit sales',
        };
      }
      payloadMeta.offline_credit_cached_ack = true;
      payloadMeta.offline_payment_note = OFFLINE_CREDIT_CACHED_NOTE;
      continue;
    }

    if (OFFLINE_PROVIDER_METHODS.has(method) || !OFFLINE_CASH_METHODS.has(method)) {
      if (!isSupervisor) {
        return {
          payloadMeta,
          blocked: true,
          error: `Offline ${method} payments are blocked for cashiers — use cash or wait until online`,
        };
      }
      const reason = (opts?.supervisorReason || '').trim();
      if (reason.length < 3) {
        return {
          payloadMeta,
          blocked: true,
          requiresSupervisorPrompt: true,
          error: `Supervisor acknowledgment required for offline ${method} (provider not verified)`,
        };
      }
      payloadMeta.offline_supervisor_ack = true;
      payloadMeta.offline_supervisor_reason = reason;
      payloadMeta.offline_provider_pending_verification = true;
      payloadMeta.offline_payment_note =
        'Provider payment queued offline — pending verification on sync (not provider-approved)';
    }
  }

  return {
    payloadMeta,
    userMessage: payloadMeta.offline_payment_note
      ? String(payloadMeta.offline_payment_note)
      : undefined,
  };
}

export function isOfflineProviderMethod(method: string): boolean {
  const norm = normalizePaymentMethod(method);
  return OFFLINE_PROVIDER_METHODS.has(norm);
}

export function isOfflineCheckout(online: boolean): boolean {
  return typeof navigator !== 'undefined' && (!online || !navigator.onLine);
}
